### PR TITLE
Fix for healing while having a damage buff

### PR DESCRIPTION
**Pull request by Paper-2**
# Description

separates healing and damage to their own modifiers.

## Related issue(s)

https://github.com/Saplings-Projects/1M_sub/issues/73

## List of changes

- a new modifier named HEAL
- a new effect to heal 

## Tests

The test in test_health.gd checks if the heal is correctly applied when having the damage buffed.

### CODE REVIEW

**Review by Turtyo — CHANGES_REQUESTED** [review]
I think it would be nice to add a new status that buffs heal and an associated card ! Just to be sure everything works correctly
Good job on the fix, I think it will work as it is what i would have done myself

**Review by Turtyo — COMMENTED** [review]
Waiting for the card that buffs heal and the related test. I'll approve once that's done !

File: Status/Debuffs/Debuff_healing.gd [hunk 1]
```diff
@@ -0,0 +1,16 @@
+extends DebuffBase
+class_name Debuff_Healing
+## reduces the target's ability to heal itself
+
+## @Override
+func _init() -> void:
+	is_on_turn_start = true
```
Comment (Turtyo) [inline Status/Debuffs/Debuff_healing.gd:7]: this is not an `on_turn_start`, this is a `is_stat_modification = true`

File: Status/Debuffs/Debuff_healing.gd [hunk 2]
```diff
@@ -0,0 +1,16 @@
+extends DebuffBase
+class_name Debuff_Healing
```
Comment (Turtyo) [inline Status/Debuffs/Debuff_healing.gd:2]: class name is `Debuff_Healing` but filename is `Debuff_**h**ealing` which is not consistent

File: Cards/Resource/Card_Debuff_healing.tres [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+value = 0
+targeting_function = SubResource("Resource_7ilgy")
+
+[sub_resource type="Resource" id="Resource_mrt6b"]
+script = ExtResource("6_augqn")
+energy_cost = 0
+
+[resource]
+script = ExtResource("5_8c6ak")
+application_type = 0
```
Comment (Turtyo) [inline Cards/Resource/Card_Debuff_healing.tres:44]: 0 is friendly only meaning we can't debuff the healing of enemies, this should be set to `All` or `Enemy only`

File: Cards/Resource/Card_Debuff_healing.tres [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+card_title = "debuff healing"
+card_description = "Debuffs you capacity to heal by 1"
```
Comment (Turtyo) [inline Cards/Resource/Card_Debuff_healing.tres:46]: Debuff the capacity of the target to heal by 1 for 3 turns

File: Cards/Resource/Card_Buff_healing.tres [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+[sub_resource type="Resource" id="Resource_308lg"]
+script = ExtResource("4_wyd1i")
+energy_cost = 0
+
+[resource]
+script = ExtResource("5_wu6tg")
+application_type = 0
+card_title = "Buff heal"
+card_description = "Buffs healing by 1"
```
Comment (Turtyo) [inline Cards/Resource/Card_Buff_healing.tres:46]: by 1 for 3 turns

**Review by Turtyo — CHANGES_REQUESTED** [review]

File: Cards/Resource/Card_Debuff_healing.tres [hunk 6]
```diff
@@ -0,0 +1,48 @@
+[gd_resource type="Resource" script_class="CardBase" load_steps=14 format=3 uid="uid://bek3vbtsf3x2l"]
+
+[ext_resource type="Script" path="res://Cards/Effects/EffectApplyStatus.gd" id="1_63tqs"]
+[ext_resource type="Script" path="res://Status/Debuffs/Debuff_healing.gd" id="2_7i6kf"]
+[ext_resource type="Script" path="res://Entity/Components/Stats/StatModifiers.gd" id="3_omw4j"]
+[ext_resource type="Script" path="res://Cards/Effects/EffectData.gd" id="4_hb64u"]
+[ext_resource type="Script" path="res://Cards/CardBase.gd" id="5_8c6ak"]
+[ext_resource type="Script" path="res://Cards/Targeting/TargetPlayer.gd" id="5_u0kut"]
```
Comment (Turtyo) [inline Cards/Resource/Card_Debuff_healing.tres:8]: if the card is targeting all then the targeting function shouldn't be target player but target base

**Review by Turtyo — APPROVED** [review]

File: Cards/Resource/Card_Debuff_healing.tres [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+[sub_resource type="Resource" id="Resource_mrt6b"]
+script = ExtResource("6_augqn")
+energy_cost = 0
+
+[resource]
+script = ExtResource("5_8c6ak")
+application_type = 2
+card_title = "debuff healing"
+card_description = "Debuffs you capacity to heal by 1 for 3 turns"
```
Comment (Tysterman74) [inline Cards/Resource/Card_Debuff_healing.tres:46]: Grammar here "Debuffs your capacity to heal by 1 for 3 turns"

File: Cards/Effects/EffectHeal.gd [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+## @Override [br]
+## Refer to [EffectBase]
+func apply_effect(caster: Entity, target: Entity, value: int) -> void:
+
+	# calculate modified heal given caster and target stats
+	var heal: float = EntityStats.get_value_modified_by_stats(GlobalEnums.PossibleModifierNames.HEAL, caster, target, value)
+	
+	# healing is essentially negative damage so we pass the var heal as negative value to make it "heal" the target
+	target.get_health_component().deal_damage(-heal, caster)
```
Comment (Tysterman74) [inline Cards/Effects/EffectHeal.gd:15]: Kind of a nitpick here, but do we want to multiply the value calculated for the heal by -1 to make it a bit more readable? I feel adding `-` in that deal_damage may get missed.

Or better yet, we can make another function in the health component for heal_damage and do the negative then? 

```
var HEAL_MODIFIER = -1
var heal: float = EntityStats.get_value_modified_by_stats(GlobalEnums.PossibleModifierNames.HEAL, caster, target, value) * HEAL_MODIFIER
```

**Review by Tysterman74 — CHANGES_REQUESTED** [review]

File: Cards/Effects/EffectHeal.gd (hunk 8 above)
Comment (Paper-2) [reply Cards/Effects/EffectHeal.gd:15]: My original plan was to have two different functions. Now that you mentioned it again, I wonder what should happen if due to a big number of negative modifiers, (For example, I want to heal 1 but I have like -10 in the healing stat) Should the player heal 0? Or should it cause damage?

**Review by Tysterman74 — APPROVED** [review]

File: Entity/Components/HealthComponent.gd [hunk 9]
```diff
@@ -20,27 +20,33 @@ func _ready() -> void:
...
+func modify_health(amount: float, caster: Entity, is_healing: bool = false) -> void:
 	# Allow caster to be null, but not the target.
-	# If caster is null, we assume that the damage came from an unknown source,
+	# If caster is null, we assume that the modification came from an unknown source,
 	# so status won't calculate.
-	
-	if damage == 0.0:
+	var new_health: float
+
+	if amount == 0.0:
```
[the remark this answers is not in the document]
Comment (Paper-2) [reply Entity/Components/HealthComponent.gd:33]: I swear I did that, lol. Dementia 🤦

**Review by Tysterman74 — APPROVED** [review]

**Pull request merged**